src/parser.cpp

```diff
@@ -64,25 +60,22 @@ namespace ada::parser {
     // Keep running the following state machine by switching on state.
     // If after a run pointer points to the EOF code point, go to the next step.
     // Otherwise, increase pointer by 1 and continue with the state machine.
-    for (; input_position <= input_size; input_position++) {
+    while(input_position <= input_size) {
```

Comment:
Can you add the `assert` @ronag recommended?

Reply:
What I think we have now is a *forward only* process. The variable input_position can only get larger. (Except in case of overflows... which could happen if the input_size exceeds the 32-bit limit on a 32-bit processor... but if you have such large strings, you have other problems.)